haigha/transports/socket_transport.py

```diff
@@ -29,26 +34,33 @@ def connect(self, (host, port), klass=socket.socket):
         Connect assuming a host and port tuple.
         '''
         self._host = "%s:%s" % (host, port)
-        self._sock = klass()
-        self._sock.setblocking(True)
-        self._sock.settimeout(self.connection._connect_timeout)
-        if self.connection._sock_opts:
-            for k, v in self.connection._sock_opts.iteritems():
-                family, type = k
-                self._sock.setsockopt(family, type, v)
-        infos = socket.getaddrinfo(host, port, 0, 0, socket.IPPROTO_TCP)
-        exc = socket.error("getaddrinfo returns an empty list")
-        for _family, _socktype, _proto, _canonname, sockaddr in infos:
+
+        for info in socket.getaddrinfo(host, port, 0, 0, socket.IPPROTO_TCP):
+
+            family, socktype, proto, _, sockaddr = info
+            self._sock = klass(family, socktype, proto)
+            self._sock.setblocking(True)
+            self._sock.settimeout(self.connection._connect_timeout)
+            if self.connection._sock_opts:
+                for k, v in self.connection._sock_opts.iteritems():
+                    family, type = k
+                    self._sock.setsockopt(family, type, v)
             try:
+
                 self._sock.connect(sockaddr)
-                break
-            except socket.error as exc:
+
+            except Exception:
+
+                LOG.exception("Failed to connect to %s:", sockaddr)
                 continue
+
+            # After connecting, switch to full-blocking mode.
+            self._sock.settimeout(None)
+            break
+
         else:
-            raise exc
 
-        # After connecting, switch to full-blocking mode.
-        self._sock.settimeout(None)
+            raise ConnectionError("No valid socket object could be created.")
```

Comment:
If getaddrinfo doesn't find anything, it raises `socket.gaierror`, so we shouldn't have to worry about an empty list.